deltachat-ffi/src/providers.rs

```diff
@@ -0,0 +1,93 @@
+use deltachat::dc_tools::{as_str, dc_strdup, StrExt};
+extern crate deltachat_provider_overview;
+use deltachat_provider_overview::StatusState;
+use std::ptr;
+
+#[no_mangle]
+pub type dc_provider = deltachat_provider_overview::Provider;
+
+#[no_mangle]
+pub unsafe extern "C" fn dc_get_json_provider_info_from_email(
+    email: *const libc::c_char,
+) -> *mut libc::c_char {
+    let option = deltachat_provider_overview::get_provider_info(
+        deltachat_provider_overview::get_domain_from_email(as_str(email)),
+    );
+    if let Some(provider) = option {
+        serde_json::to_string(provider.0).unwrap().strdup()
+    } else {
+        dc_strdup(ptr::null())
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn dc_get_provider_info_from_email(
+    email: *const libc::c_char,
+) -> *const dc_provider {
+    let option = deltachat_provider_overview::get_provider_info(
+        deltachat_provider_overview::get_domain_from_email(as_str(email)),
+    );
+    if let Some(provider) = option {
+        provider.0
+    } else {
+        ptr::null()
+    }
+}
+
+macro_rules! null_guard {
+    ($context:tt) => {
+        if $context.is_null() {
+            return dc_strdup(ptr::null_mut());
```

Comment:
Depending on what the result of an earlier comment is this is likely the wrong choice again since NULL is probably a better return value in these cases.

Reply:
Again I don't know if you can return null when it expects a pointer to a string.

Reply:
Every pointer can be null in C (unfortunately). @Simon-Laux Why is this ptr::null **_mut** ()? (what's the difference to ptr::null()?)

Reply:
Rust keeps different types for mutable and constant objects, even for null pointers.  So if the return type is `-> *mut libc::c_char` than the error return is `ptr::null_mut()`.  If the return type is `-> *const libc::c_char` than the error return is `ptr::null()`.

Reply:
Which makes me think it's easier to not have the macro, it doesn't save that much and you have to cross check the types.  but up to you.

Reply:
But what else should the FFI do if it is passed a null pointer? Should it just try to dereference it and panic?

Reply:
No, sorry that is not what I meant.  The rust code should never panic and if it gets passed NULL it should return an error response, usually NULL itself.

Reply:
Done.